app/models/concerns/orchestration/dns.rb

```diff
@@ -2,179 +2,224 @@ module Orchestration::DNS
     # host.managed? and managed? should always come first so that orchestration doesn't
     # even get tested for such objects
-    (host.nil? || host.managed?) && managed? && hostname.present? && ip_available? && !domain.nil? && !domain.proxy.nil? && SETTINGS[:unattended]
+    SETTINGS[:unattended] && (host.nil? || host.managed?) && managed? && hostname.present?
+  end
+
+  def dns?
+    dns_ready? && ip_available? && !domain.nil? && !domain.proxy.nil?
+  end
+
+  def dns6?
+    dns_ready? && ip6_available? && !domain.nil? && !domain.proxy.nil?
   end
 
   def reverse_dns?
-    # host.managed? and managed? should always come first so that orchestration doesn't
-    # even get tested for such objects
-    (host.nil? || host.managed?) && managed? && hostname.present? && ip_available? && !subnet.nil? && subnet.dns? && SETTINGS[:unattended]
+    dns_ready? && ip_available? && !subnet.nil? && subnet.dns?
+  end
+
+  def reverse_dns6?
+    dns_ready? && ip6_available? && !subnet6.nil? && subnet6.dns?
   end
 
   def rebuild_dns
-    logger.info "DNS not supported for #{name}, #{ip}, skipping orchestration rebuild" unless dns?
-    logger.info "Reverse DNS not supported for #{name}, #{ip}, skipping orchestration rebuild" unless reverse_dns?
-    return true unless dns? || reverse_dns?
+    logger.info "IPv4 DNS not supported for #{name}, skipping orchestration rebuild" unless dns?
+    logger.info "IPv6 DNS not supported for #{name}, skipping orchestration rebuild" unless dns6?
+    logger.info "Reverse IPv4 DNS not supported for #{name}, skipping orchestration rebuild" unless reverse_dns?
+    logger.info "Reverse IPv6 DNS not supported for #{name}, skipping orchestration rebuild" unless reverse_dns6?
+    return true unless dns? || dns6? || reverse_dns? || reverse_dns6?
     del_dns_a_record_safe
+    del_dns_aaaa_record_safe
     del_dns_ptr_record_safe
-    a_record_result, ptr_record_result = true, true
+    del_dns_ptr6_record_safe
+    a_record_result = true
+    aaaa_record_result = true
+    ptr_record_result = true
+    ptr6_record_result = true
     begin
       a_record_result = recreate_a_record if dns?
+      aaaa_record_result = recreate_aaaa_record if dns6?
       ptr_record_result = recreate_ptr_record if reverse_dns?
-      a_record_result && ptr_record_result
+      ptr6_record_result = recreate_ptr6_record if reverse_dns6?
+      a_record_result && aaaa_record_result && ptr_record_result && ptr6_record_result
     rescue => e
-      Foreman::Logging.exception "Failed to rebuild DNS record for #{name}, #{ip}", e, :level => :error
+      Foreman::Logging.exception "Failed to rebuild DNS record for #{name}, #{ip}, #{ip6}", e, :level => :error
       false
     end
   end
 
   def dns_a_record
     return unless dns? or @dns_a_record
     handle_validation_errors do
-      @dns_a_record ||= Net::DNS::ARecord.new dns_record_attrs
+      @dns_a_record ||= Net::DNS::ARecord.new dns_a_record_attrs
+    end
+  end
+
+  def dns_aaaa_record
```

Comment:
It seems like the creation/deletion parts of this class have outgrown the responsibility of this class (just maintaining the validations/callbacks to create/delete DNS records during orchestration. I think it'd be best to move creation and deletion to their own modules like:

```ruby 
module DNSRecordCreation
  def create_dns_record(type)
    return if instance_variable_get("@dns_#{type}_record").present?
    case type
    when 'aaaa', 'ptr6'
      return unless dns6? && ip6_available?
    when 'a', 'ptr' 
      return unless dns? && ip_available?
    end

    handle_validation_errors do
      instance_variable_set(
        "@dns_#{type}_record", 
        dns_class(type).new(send(:"dns_#{type}_record_attrs"))
    end
  end
  
  private
 
  def dns_class(type)
    class_name = type.upcase.constantize
    Net::DNS::class_name
  end
end
```

Creation and deletion of DNS entries would go like `create_dns_record('aaaa')` and it seems like at least each file would have its own responsibility.